docs/docker.md

```diff
@@ -0,0 +1,108 @@
+# Docker inside Envbuilder
+
+There are a number of approaches you can use to have access to a Docker daemon
+from inside Envbuilder:
+
+## Docker Outside of Docker (DooD)
+
+**Security:** None
+**Convenience:** High
+
+This approach re-uses the host Docker socket and passes it inside the container.
+It is the simplest approach, but offers **no security** -- any process inside the
+container that can connect to the Docker socket will have access to the
+underlying host.
+Only use it if you are the only person using the Docker socket (for example, if
+you are experimenting on your own workstation).
+
+Example:
+
+```shell
+docker run -it --rm \
+    -v /tmp/envbuilder:/workspaces \
+    -e ENVBUILDER_GIT_URL=https://github.com/coder/envbuilder \
+    -e ENVBUILDER_DEVCONTAINER_DIR=/workspaces/envbuilder/examples/docker/01_dood \
+    -e ENVBUILDER_INIT_SCRIPT=bash \
+    ghcr.io/coder/envbuilder:latest
+```
+
+
+## Docker-in-Docker (DinD)
+
+**Security:** Low
+**Convenience:** High
+
+This approach entails running a Docker daemon inside the container.
+This requires a privileged container to run, and therefore has a wide potential
+attack surface.
+
+Example:
+
+> Note that due to a lack of init system, the Docker daemon
+> needs to be started separately inside the container. In this example, we
+> create a custom entrypoint to start the Docker daemon in the background and
+> call this entrypoint via `ENVBUILDER_INIT_SCRIPT`.
+
+```shell
+docker run -it --rm \
+    --privileged \
+    -v /tmp/envbuilder:/workspaces \
+    -e ENVBUILDER_GIT_URL=https://github.com/coder/envbuilder \
+    -e ENVBUILDER_DEVCONTAINER_DIR=/workspaces/envbuilder/examples/docker/02_dind \
+    -e ENVBUILDER_INIT_SCRIPT=/entrypoint.sh \
+    ghcr.io/coder/envbuilder:latest
+```
+
+This can also be accomplished using the [`docker-in-docker` Devcontainer feature](https://github.com/devcontainers/features/tree/main/src/docker-in-docker).
```

Comment:
Just off the top of my head, this won't work OOTB, envbuilder doesn't run the entrypoint defined here:

https://github.com/devcontainers/features/blob/67c10a660868260c284b02f2878dcfdcfd91279f/src/docker-in-docker/devcontainer-feature.json#L60

Mounts may also require special consideration. 🤔

Reply:
Yeah, the custom entrypoint is still required here. 👍

Reply:
I added a separate example `03_dind_feature` for this. Works fine with volume mounts as far as I tested, too.